### ISSUE TITLE
Integrate SLATE Remote Client auto-upgrade with GitHub Releases

### ISSUE THREAD

**Issue by adamhgriffith-uofu** [slateci/slate-client-server#207]
The SLATE Remote Client currently relies on [latest.json](https://jenkins.slateci.io/artifacts/client/latest.json) to determine if/when an update is available. Our new paradigm will involve GitHub Releases where a URL like https://github.com/slateci/slate-client-server/releases/latest will take you to the page for the latest release. We need to update the code to interact with how GitHub does things.

Adding @adamhgriffith-uofu as a watcher.

**Comment by cnweaver** [comment]
It should be noted that https://jenkins.slateci.io/artifacts/client/latest.json is designed to not only provide indirection to the actual download URLs, but to provide an efficient, machine-readable means to check whether a newer version exists. The proposed GitHub mechanism breaks this intelligent update capability, unless fragile HTML scraping is employed to extract the version number from the resulting document after the GitHub redirect. In addition, more complex logic will be required to compare version numbers. This seems like considerable effort to little purpose.

**Comment by cnweaver** [comment]
I think that doing that, and having some stable URL at which it is then stored should solve the main problem; that's basically what the Jenkins build was set up to do.

**Comment by adamhgriffith-uofu** [comment]
After rummaging around I just found GitHub's pretty nice [REST API](https://docs.github.com/en/rest/releases/releases) for releases these days. E.g.
```shell
$ curl https://api.github.com/repos/slateci/slate-client-server/releases/latest
```
response:
```yaml
{
  "url": "https://api.github.com/repos/slateci/slate-client-server/releases/67719578",
  "assets_url": "https://api.github.com/repos/slateci/slate-client-server/releases/67719578/assets",
  "upload_url": "https://uploads.github.com/repos/slateci/slate-client-server/releases/67719578/assets{?name,label}",
  "html_url": "https://github.com/slateci/slate-client-server/releases/tag/v0.0.3",
  "id": 67719578,
  "author": {
    "login": "github-actions[bot]",
    "id": 41898282,
    "node_id": "MDM6Qm90NDE4OTgyODI=",
    "avatar_url": "https://avatars.githubusercontent.com/in/15368?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/github-actions%5Bbot%5D",
    "html_url": "https://github.com/apps/github-actions",
    "followers_url": "https://api.github.com/users/github-actions%5Bbot%5D/followers",
    "following_url": "https://api.github.com/users/github-actions%5Bbot%5D/following{/other_user}",
    "gists_url": "https://api.github.com/users/github-actions%5Bbot%5D/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/github-actions%5Bbot%5D/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/github-actions%5Bbot%5D/subscriptions",
    "organizations_url": "https://api.github.com/users/github-actions%5Bbot%5D/orgs",
    "repos_url": "https://api.github.com/users/github-actions%5Bbot%5D/repos",
    "events_url": "https://api.github.com/users/github-actions%5Bbot%5D/events{/privacy}",
    "received_events_url": "https://api.github.com/users/github-actions%5Bbot%5D/received_events",
    "type": "Bot",
    "site_admin": false
  },
  "node_id": "RE_kwDOCgZjeM4ECVGa",
  "tag_name": "v0.0.3",
  "target_commitish": "fd58966d36ef0e62fc5640de350bda3af71569fd",
  "name": "Release v0.0.3",
  "draft": false,
  "prerelease": false,
  "created_at": "2022-05-24T17:22:04Z",
  "published_at": "2022-05-24T17:28:30Z",
  "assets": [
    {
      "url": "https://api.github.com/repos/slateci/slate-client-server/releases/assets/66498859",
      "id": 66498859,
      "node_id": "RA_kwDOCgZjeM4D9rEr",
      "name": "slate-linux.sha256",
      "label": "",
      "uploader": {
        "login": "github-actions[bot]",
        "id": 41898282,
        "node_id": "MDM6Qm90NDE4OTgyODI=",
        "avatar_url": "https://avatars.githubusercontent.com/in/15368?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/github-actions%5Bbot%5D",
        "html_url": "https://github.com/apps/github-actions",
        "followers_url": "https://api.github.com/users/github-actions%5Bbot%5D/followers",
        "following_url": "https://api.github.com/users/github-actions%5Bbot%5D/following{/other_user}",
        "gists_url": "https://api.github.com/users/github-actions%5Bbot%5D/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/github-actions%5Bbot%5D/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/github-actions%5Bbot%5D/subscriptions",
        "organizations_url": "https://api.github.com/users/github-actions%5Bbot%5D/orgs",
        "repos_url": "https://api.github.com/users/github-actions%5Bbot%5D/repos",
        "events_url": "https://api.github.com/users/github-actions%5Bbot%5D/events{/privacy}",
        "received_events_url": "https://api.github.com/users/github-actions%5Bbot%5D/received_events",
        "type": "Bot",
        "site_admin": false
      },
      "content_type": "text/plain",
      "state": "uploaded",
      "size": 85,
      "download_count": 0,
      "created_at": "2022-05-24T17:28:32Z",
      "updated_at": "2022-05-24T17:28:32Z",
      "browser_download_url": "https://github.com/slateci/slate-client-server/releases/download/v0.0.3/slate-linux.sha256"
    },
    {
      "url": "https://api.github.com/repos/slateci/slate-client-server/releases/assets/66498858",
      "id": 66498858,
      "node_id": "RA_kwDOCgZjeM4D9rEq",
      "name": "slate-linux.tar.gz",
      "label": "",
      "uploader": {
        "login": "github-actions[bot]",
        "id": 41898282,
        "node_id": "MDM6Qm90NDE4OTgyODI=",
        "avatar_url": "https://avatars.githubusercontent.com/in/15368?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/github-actions%5Bbot%5D",
        "html_url": "https://github.com/apps/github-actions",
        "followers_url": "https://api.github.com/users/github-actions%5Bbot%5D/followers",
        "following_url": "https://api.github.com/users/github-actions%5Bbot%5D/following{/other_user}",
        "gists_url": "https://api.github.com/users/github-actions%5Bbot%5D/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/github-actions%5Bbot%5D/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/github-actions%5Bbot%5D/subscriptions",
        "organizations_url": "https://api.github.com/users/github-actions%5Bbot%5D/orgs",
        "repos_url": "https://api.github.com/users/github-actions%5Bbot%5D/repos",
        "events_url": "https://api.github.com/users/github-actions%5Bbot%5D/events{/privacy}",
        "received_events_url": "https://api.github.com/users/github-actions%5Bbot%5D/received_events",
        "type": "Bot",
        "site_admin": false
      },
      "content_type": "application/gzip",
      "state": "uploaded",
      "size": 311369,
      "download_count": 0,
      "created_at": "2022-05-24T17:28:31Z",
      "updated_at": "2022-05-24T17:28:31Z",
      "browser_download_url": "https://github.com/slateci/slate-client-server/releases/download/v0.0.3/slate-linux.tar.gz"
    },
    {
      "url": "https://api.github.com/repos/slateci/slate-client-server/releases/assets/66498861",
      "id": 66498861,
      "node_id": "RA_kwDOCgZjeM4D9rEt",
      "name": "slate-macos.sha256",
      "label": "",
      "uploader": {
        "login": "github-actions[bot]",
        "id": 41898282,
        "node_id": "MDM6Qm90NDE4OTgyODI=",
        "avatar_url": "https://avatars.githubusercontent.com/in/15368?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/github-actions%5Bbot%5D",
        "html_url": "https://github.com/apps/github-actions",
        "followers_url": "https://api.github.com/users/github-actions%5Bbot%5D/followers",
        "following_url": "https://api.github.com/users/github-actions%5Bbot%5D/following{/other_user}",
        "gists_url": "https://api.github.com/users/github-actions%5Bbot%5D/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/github-actions%5Bbot%5D/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/github-actions%5Bbot%5D/subscriptions",
        "organizations_url": "https://api.github.com/users/github-actions%5Bbot%5D/orgs",
        "repos_url": "https://api.github.com/users/github-actions%5Bbot%5D/repos",
        "events_url": "https://api.github.com/users/github-actions%5Bbot%5D/events{/privacy}",
        "received_events_url": "https://api.github.com/users/github-actions%5Bbot%5D/received_events",
        "type": "Bot",
        "site_admin": false
      },
      "content_type": "text/plain",
      "state": "uploaded",
      "size": 85,
      "download_count": 0,
      "created_at": "2022-05-24T17:28:33Z",
      "updated_at": "2022-05-24T17:28:33Z",
      "browser_download_url": "https://github.com/slateci/slate-client-server/releases/download/v0.0.3/slate-macos.sha256"
    },
    {
      "url": "https://api.github.com/repos/slateci/slate-client-server/releases/assets/66498860",
      "id": 66498860,
      "node_id": "RA_kwDOCgZjeM4D9rEs",
      "name": "slate-macos.tar.gz",
      "label": "",
      "uploader": {
        "login": "github-actions[bot]",
        "id": 41898282,
        "node_id": "MDM6Qm90NDE4OTgyODI=",
        "avatar_url": "https://avatars.githubusercontent.com/in/15368?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/github-actions%5Bbot%5D",
        "html_url": "https://github.com/apps/github-actions",
        "followers_url": "https://api.github.com/users/github-actions%5Bbot%5D/followers",
        "following_url": "https://api.github.com/users/github-actions%5Bbot%5D/following{/other_user}",
        "gists_url": "https://api.github.com/users/github-actions%5Bbot%5D/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/github-actions%5Bbot%5D/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/github-actions%5Bbot%5D/subscriptions",
        "organizations_url": "https://api.github.com/users/github-actions%5Bbot%5D/orgs",
        "repos_url": "https://api.github.com/users/github-actions%5Bbot%5D/repos",
        "events_url": "https://api.github.com/users/github-actions%5Bbot%5D/events{/privacy}",
        "received_events_url": "https://api.github.com/users/github-actions%5Bbot%5D/received_events",
        "type": "Bot",
        "site_admin": false
      },
      "content_type": "application/gzip",
      "state": "uploaded",
      "size": 362981,
      "download_count": 0,
      "created_at": "2022-05-24T17:28:32Z",
      "updated_at": "2022-05-24T17:28:33Z",
      "browser_download_url": "https://github.com/slateci/slate-client-server/releases/download/v0.0.3/slate-macos.tar.gz"
    }
  ],
  "tarball_url": "https://api.github.com/repos/slateci/slate-client-server/tarball/v0.0.3",
  "zipball_url": "https://api.github.com/repos/slateci/slate-client-server/zipball/v0.0.3",
  "body": "See https://slateci.io/docs/tools/ for installation and usage instructions."
}
```

**Comment by cnweaver** [comment]
That looks much more suitable than the human-facing web page. Other considerations to look at are whether it's a good idea to change the version numbering system (e.g. doing so will render all currently installed clients un-updatable), and how to have the client figure out the mapping from its platform name to the right download URL in the GitHub JSON. There's a `label` field, which looks empty at the moment; if it's possible to fill that in with the platform names that would work nicely, although pattern matching for `name == slate-${platform}.tar.*` would work if necessary.

**Comment by adamhgriffith-uofu** [comment]
Those are good points we will need to consider as we work on this. I did try to populate `label` but it does not appear currently possible using the [ncipollo/release-action](https://github.com/ncipollo/release-action) GitHub action.

**Comment by adamhgriffith-uofu** [comment]
After doing some reading for a parallel task we can grab the latest client with the following simple command:
```shell
curl -LO https://github.com/slateci/slate-client-server/releases/latest/download/slate-linux.tar.gz
```

**Comment by adamhgriffith-uofu** [comment]
Adding comment in Slack from @LincolnBryant:
"slate version upgrade actually downgrades your client :exploding_head:"
```
[14:58] uct2-int.mwt2.org:~ $ slate version
Client Version Server Version
1.0.31         1.0.31        
Server supported API versions: v1alpha3
[14:58] uct2-int.mwt2.org:~ $ slate version upgrade
Client Version Server Version
1.0.31         1.0.31        
Server supported API versions: v1alpha3
Version 1908 is available; this executable is version 1.0.31
Do you want to download and install the new version? [Y/n] y
Upgraded to version 1908
[14:58] uct2-int.mwt2.org:~ $ slate version
Client Version Server Version
1908           1.0.31        
Server supported API versions: v1alpha3
```
